### PR TITLE
Invoice

### PR DESCRIPTION
**Pull request by guillaime**
Merge request for invoice to dev

### CODE REVIEW

File: src/main/java/dao/InvoiceDao.kt [hunk 1]
```diff
@@ -45,11 +45,10 @@ class InvoiceDao {
             .setParameter("state", state)
             .resultList
 
-    fun getAllInvoicesByProfile(profileId: Long?): List<Invoice> {
-        val query = em!!.createQuery("SELECT i FROM Invoice i " + "JOIN i.profile c WHERE c.id = :id", Invoice::class.java)
-
-        return query.setParameter("id", profileId).resultList
-    }
+    fun allInvoicesByCivilian(civilianId: String): List<Invoice> = em
```
Comment (Requinard) [inline src/main/java/dao/InvoiceDao.kt]: Remove the type annotation, eg ` fun allInvoicesByCivilian(civilianId: String) = em`

File: src/main/java/dao/InvoiceDao.kt [hunk 2]
```diff
@@ -45,11 +45,10 @@ class InvoiceDao {
...
+            .createQuery("SELECT i FROM Invoice i JOIN i.profile p WHERE p.id = :id", Invoice::class.java)
```
Comment (Requinard) [inline src/main/java/dao/InvoiceDao.kt]: Aren't invoices linked to a profile? That way you should be able to get them without using a query

File: src/main/java/domain/Activity.kt [hunk 3]
```diff
@@ -2,13 +2,17 @@ package domain
...
 import java.io.Serializable
+import java.util.*
 
 @Entity
 class Activity : Serializable {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    private val id: Long? = null
+    private val id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Activity.kt]: Wasn't this  supposed to be moved to hibernate to properly generate this?

File: src/main/java/domain/Activity.kt [hunk 4]
```diff
@@ -2,13 +2,17 @@ package domain
...
+
+    constructor(){
```
Comment (Requinard) [inline src/main/java/domain/Activity.kt]: Remove this empty constructor as this is automatically done with gradle

File: src/main/java/domain/Civilian.kt [hunk 5]
```diff
@@ -2,14 +2,14 @@ package domain
...
 @Entity
 @DiscriminatorValue("Civilian")
 class Civilian : KontoUser(), Serializable {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    var id: Long? = null
+    var id: String? = UUID.randomUUID().toString()
 
     @ManyToMany
```
Comment (Requinard) [inline src/main/java/domain/Civilian.kt]: Where is the mapping query?

File: src/main/java/domain/Invoice.kt [hunk 6]
```diff
@@ -5,24 +5,23 @@ import domain.enums.InvoiceState
...
+    var id: String? = UUID.randomUUID().toString()
     @Column
     var createdOn: Date? = null
     @Column
     var generatedFor: Date? = null
     @Enumerated(EnumType.STRING)
     var generationType: InvoiceGenerationType? = null
     @Enumerated(EnumType.STRING)
-    var state: InvoiceState? = null
+    var state: InvoiceState? = InvoiceState.OPEN
```
Comment (Requinard) [inline src/main/java/domain/Invoice.kt]: If the invoice status defaults to OPEN, it should not be nullable

File: src/main/java/domain/KontoUser.kt [hunk 7]
```diff
@@ -1,15 +1,15 @@
 package domain
 
+import java.util.*
 import javax.persistence.*
 
 @Entity
-@Table(name = "KontoUser")
+@Table(name = "KONTOUSER")
```
Comment (Requinard) [inline src/main/java/domain/KontoUser.kt]: for SQL it's default to capitalize keywords as `WHERE, LIKE and SELECT` and keep table names to lowercase

File: src/main/java/domain/KontoUser.kt [hunk 8]
```diff
@@ -1,15 +1,15 @@
...
 @Inheritance(strategy = InheritanceType.JOINED)
 abstract class KontoUser {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    private val id: Long? = null
+    private val id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/KontoUser.kt]: If there is a default parameter it should not be nullable

File: src/main/java/domain/Location.kt [hunk 9]
```diff
@@ -1,13 +1,13 @@
...
+import java.util.*
 import javax.persistence.*
 
 @Entity
 class Location {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    private val id: Long? = null
+    private val id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Location.kt]: same

File: src/main/java/domain/Profile.kt [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 
 import javax.persistence.*
 
 @Entity
 class Profile {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    var id: Long? = null
+    var id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Profile.kt]: same

File: src/main/java/domain/Rate.kt [hunk 11]
```diff
@@ -11,7 +11,7 @@ import java.util.*
 class Rate : Serializable {
 
     @Id
-    var id: String? = null
+    var id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Rate.kt]: same

File: src/main/java/domain/Vehicle.kt [hunk 12]
```diff
@@ -4,15 +4,14 @@ import domain.enums.VehicleType
 
 import javax.persistence.*
 import java.io.Serializable
-import java.util.ArrayList
+import java.util.*
 
 @Entity
-@Table(name = "Vehicle")
+@Table(name = "VEHICLE")
```
Comment (Requinard) [inline src/main/java/domain/Vehicle.kt]: same

File: src/main/java/domain/Vehicle.kt [hunk 13]
```diff
@@ -4,15 +4,14 @@ import domain.enums.VehicleType
...
 class Vehicle : Serializable {
 
     @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    var id: Long? = null
+    var id: String? = UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Vehicle.kt]: idem

File: src/main/java/domain/Vignette.kt [hunk 14]
```diff
@@ -11,7 +11,7 @@ import java.util.*
 class Vignette : Serializable {
 
     @Id
-    var id: String ?= null
+    var id: String ?= UUID.randomUUID().toString()
```
Comment (Requinard) [inline src/main/java/domain/Vignette.kt]: same

File: src/main/java/rest/InvoiceResource.kt [hunk 15]
```diff
@@ -66,4 +63,11 @@ class InvoiceResource @Inject constructor(
     @Produces("application/json")
     fun allInvoicesByState(@PathParam("state") state: String): List<Invoice> =
             invoiceService.allInvoicesByState(InvoiceState.valueOf(state))
+
+    @POST
+    @Path("/state/change/")
+    @Produces("application/json")
+    fun updateInvoiceState(@FormParam("invoiceId") invoiceId: String,
```
Comment (Requinard) [inline src/main/java/rest/InvoiceResource.kt]: Has this endpoint been properly tested?

File: src/main/java/dao/InvoiceDao.kt (hunk 1 above)
Comment (guillaime) [reply src/main/java/dao/InvoiceDao.kt]: Removed function

File: src/main/java/dao/InvoiceDao.kt (hunk 2 above)
Comment (guillaime) [reply src/main/java/dao/InvoiceDao.kt]: Function has been removed, invoices linked by a profile now are received by the userDao

File: src/main/java/domain/Activity.kt (hunk 3 above)
Comment (guillaime) [reply src/main/java/domain/Activity.kt]: DONE

File: src/main/java/domain/Activity.kt (hunk 4 above)
Comment (guillaime) [reply src/main/java/domain/Activity.kt]: DONE

File: src/main/java/domain/Invoice.kt (hunk 6 above)
Comment (guillaime) [reply src/main/java/domain/Invoice.kt]: DONE

File: src/main/java/domain/KontoUser.kt (hunk 7 above)
Comment (guillaime) [reply src/main/java/domain/KontoUser.kt]: DONE

File: src/main/java/domain/KontoUser.kt (hunk 8 above)
Comment (guillaime) [reply src/main/java/domain/KontoUser.kt]: DONE

File: src/main/java/domain/Location.kt (hunk 9 above)
Comment (guillaime) [reply src/main/java/domain/Location.kt]: DONE

File: src/main/java/domain/Profile.kt (hunk 10 above)
Comment (guillaime) [reply src/main/java/domain/Profile.kt]: DONE

File: src/main/java/domain/Rate.kt (hunk 11 above)
Comment (guillaime) [reply src/main/java/domain/Rate.kt]: DONE

File: src/main/java/domain/Vehicle.kt (hunk 12 above)
Comment (guillaime) [reply src/main/java/domain/Vehicle.kt]: DONE

File: src/main/java/domain/Vehicle.kt (hunk 13 above)
Comment (guillaime) [reply src/main/java/domain/Vehicle.kt]: DONE

File: src/main/java/domain/Vignette.kt (hunk 14 above)
Comment (guillaime) [reply src/main/java/domain/Vignette.kt]: DONE

File: src/main/java/rest/InvoiceResource.kt (hunk 15 above)
Comment (guillaime) [reply src/main/java/rest/InvoiceResource.kt]: Added @Consumes for now, later this will be changed. Created new JIRA [PTS6-113] Issue to do this.

**Pull request merged**